magma/lib/magma/attributes/file_collection_attribute.rb

```diff
@@ -0,0 +1,155 @@
+        serializer.to_payload_format(record_name, file_hash, loader.user, index)
+      end
+      [name, payload_format ]
+    end
+
+    def query_to_payload(data)
+      return nil unless data
+
+      data.map do |datum|
+        serializer.to_query_payload_format(datum)
+      end
+    end
+
+    def query_to_tsv(files)
+      files.map do |file|
+        serializer.to_query_tsv_format(file)
+      end
+    end
+
+    def entry(files, loader)
+      entry_format = files.map do |file|
+        serializer.to_loader_entry_format(file)
+      end.compact  # Remove the ::temp `nil` values for the loader
+
+      [ column_name, entry_format.to_json ]
+    end
+
+    def load_hook(loader, record_name, files, copy_revisions)
+      # Revising a FileCollectionAttribute makes some assumptions about the revisions.
+      # We need to make sure that the links aren't lost on the Metis
+      #   end of things, since that's the only pointer to the pre-revision data_block
+      #   that we have for each Magma file.
+      # For example,
+      #     [file[1], file[0]] is a bad revision, because on Metis's end,
+      #   after we update the new index-0 file (magma-file-0 -> original-magma-file-1-data-block),
+      #   we lose the pointer to the original magma-file-0 data_block.
+      #   It becomes impossible to do the magma-file-1 -> original-magma-file-0-data-block
+      #   revision.
+      # Thus, we must enforce these requirements:
+      #   1) Any existing files must appear in ascending index order.
+      #       [file[0], file[1], file[7], file[11]] is okay.
+      #       [file[11], file[1], file[0], file[7]] is NOT okay.
+      #   2) New files / Metis paths must appear at the end of the sequence.
+      #       [file[1], file[8], {path: "metis://foo/bar/bim.txt"}] is okay.
+      #       [{path: "metis://bad/revision/order.txt"}, file[1], file[8]] is NOT okay.
+      #   3) Any ::temp files must be at the end of the array, so the filename indices stay sequential.
+      #       [file[1], {path: "metis://foo/bar/bim.txt"}, "::temp"] is okay.
+      #       [file[1], "::temp", {path: "metis://foo/bar/bim.txt"}] is not okay, because "bim.txt" will get assigned index 2 and
+      #           they will appear to skip -- [file[0], file[2]] would be stored, instead of [file[0], file[1]].
+      # In this hook we return an error message if any of the above
+      #   requirements are not met, to stop the update.
+      # Note: We'll also have to consider that "new files" may be ::temp or ::blank.
+
+      return "Not all files have the required :path key." unless files.all? { |file| file&.dig(:path) }
+
+      return "Existing files are not in ascending index order." unless existing_files_in_sequence?(record_name, files)
+
+      return "New files must be at the end of the revision." unless new_files_at_end?(record_name, files)
+
+      return "Temp files must be at the end of the revision." unless temp_files_at_end?(record_name, files)
+
```

Comment:
I'm not sure I understand the need for all of these ordering requirements for loading, can you clarify?

Reply:
I'll do my best -- might be easier to hop on a call and chat about it. I think the first two (existing file ordering / new-files last) are hard requirements, for the following reasons.

On the Metis side of things, it processes a set of copy commands sequentially, in the order they are sent. If we start with a set of newly added files, we provide Metis paths like so:

```
[{path: "metis/file-0.txt"}, {path: "metis/file-1.txt"}]
```

Leads to a copy request like:

```
[{source: "metis/file-0.txt", dest: "magma/file-0.txt"}, {source: "metis/file-1.txt", dest: "magma/file-1.txt"}]
```

Would pass validation in Metis (assuming all files exist), and then Metis would save into the database something like:

```
File: {id: 1, folder: "magma", name: "file-0.txt", data_block: 123}
File: {id: 2, folder: "magma", name: "file-1.txt", data_block: 456}
```

Let's say you come back and make a revision where the existing-order constraint is violated, like below:

```
[{path: "magma/file-1.txt"}, {path: "magma/file-0.txt"}]
```

Note that the paths are now, by necessity, pointing to files in the Magma bucket, because that is all we have reference to. This leads to a copy request like:

```
[{source: "magma/file-1.txt", dest: "magma/file-0.txt"}, {source: "magma/file-0.txt", dest: "magma/file-1.txt"}]
```

Metis then goes through and tries to update them sequentially. So it executes the first copy:

```
File: {id: 1, folder: "magma", name: "file-0.txt", data_block: 456} <- updated first
File: {id: 2, folder: "magma", name: "file-1.txt", data_block: 456} 
```

But now when it tries to execute the second copy command, it points the file `magma/file-1.txt` to the data block pointed to by `magma/file-0.txt`, which is now`456` instead of the desired / original `123`, because the reference to data block `123` in the Magma bucket has been lost. Something similar would happen if you interspersed new files, like:

```
[{path: "metis/new-file-3.txt"}, {path: "magma/file-0.txt"}]
```

In order to make sure the data blocks are preserved for all copy requests, you have to go in ascending index order to not overwrite the data block references.

We could make Metis do a single bulk update to the database, but there isn't a "Magma::Loader" equivalent for Metis right now.

Note that there are no controls in Timur to re-order items in a ListInput, so the issue would only appear with scripts talking directly to the Magma API.

Reply:
I think the `::temp` requirement might be needed, though it may not strictly be necessary from a functional standpoint. I was thinking that you ideally want the indices to be sequential:

```
[{name: "model-record-attribute-0.txt"}, {name: "model-record-attribute-1.txt"}, {::temp}]
```

instead of 

```
[{name: "model-record-attribute-0.txt"}, {::temp}, {name: "model-record-attribute-2.txt"}]
```

Which might confuse people, who would be like "where is file -1?" (since ::temp doesn't get saved to the database).

But maybe that's not a big deal, especially since we're showing the original filename in Timur, not the generated name with the index.